### ISSUE TITLE
Feature request: Reading velocity

### ISSUE THREAD

**Issue by christophfriedrich** [react-spring/use-cannon#45]
I saw that in #31 @bvkimball asked
>  Also is it possible to get the velocity from the worker?

and @drcmda replied:
> Reading isn't possible atm and would have to be async, you'd get a promise back.

I would like to say that I'd appreciate a reading feature very much!

**Comment by nevernotsean** [comment]
Getting values in general from bodies in the worker would present a problem, but for the library to be feature complete some sort of solution for getting needs to be put into place.

Could we use an Observable pattern to solve this? Maybe leveraging RxJs which is built to handle async data streams. We could add a `subscribe` method to each property, then send updated values on each post-step?

For example:
```
const [currentVel, setCurrentVel] = useState(0)

api.velocity.subscribe((val) => setCurrentVel(val)); 

useEffect(() => {
... do stuff
}, [currentVel])
```

**Comment by codynova** [comment]
We discussed the possibility of expanding the `makeApi` function to return getters for that data, and I think that would make the most sense from an API and performance perspective.

Calling a getter would return a promise that would be resolved when the worker has returned the data to the main thread. I've started working on a PR in the last day, but am juggling a couple other work items. I should have time tonight or this weekend - but also feel free to take a stab at it if you want! This is what Paul had suggested and I think it looks good:

```js
get velocity() {
  return new Promise(resolve => {
    context.promises[id] = { op: "request", type: "velocity", resolve }
  })
```

> in the physics parent it keeps that list and if it receives a rerquest back from the worker, it resolves the promise:

```js
switch (op) {
  case "request": {
    if (promises[id]) {
      promises[id].resolve(value)
      delete promises[id]
```

**Comment by codynova** [comment]
New api which allows calling `.get()` for body props (#55) released in 0.3.0, closing

**Comment by drcmda** [comment]
sorry i didnt see the subscribe proposal @nevernotsean that also looks good. both could have ups and downs. promises: hard to use, needs constant obj creation, bad if called in rapid succession. saves mem when used sparingly. subscriptions: creates more data to go through the bridge for one-shot requests and more overhead. but better for successive requests.

**Comment by codynova** [comment]
I think that makes perfect sense. When I said "better from a performance perspective" I was imagining hooking the subscription up to state, but I should have considered more carefully.

>i am wildly guessing here, but i think one-shot requests are rare, successive requests are the norm.

I agree, especially for things like velocity, position, rotation. For boolean props, I think they are more likely to be less frequent. Does it make sense to keep the `.get()` method for one-shot request and add the `.subscribe()` method for subscriptions, or should we drop `get` entirely and go full `subscribe`?

**Comment by drcmda** [comment]
booleans, hmmmm, hard question. but i think we should choose one, for simplicity sake. i looked over your code and it's basically all there to accommodate us in either way, but we would save some bytes if we go for one and also lessen the api overhead a little.

**Comment by codynova** [comment]
Sounds good, then let's go for @nevernotsean's proposal 😃

**Comment by codynova** [comment]
How does the implementation work? Send a message from the hooks to the worker to `subscribe` to a certain property, which causes the worker to register a callback that posts a message back to the hooks on each `step`? I've only used RXJS in Angular 2 ages ago, so apologies if I'm way off base!

**Comment by drcmda** [comment]
```jsx
// subscribe
const unsub = api.velocity.subscribe(value => console.log(value))
// unsubscribe
unsub()
```

more or less like this

```jsx
const { worker, bodies, ..., listeners } = useContext(context)
...
{
  subscribe(cb) {
    listeners[id][name] = cb // name === "velocity"
    return () => delete listeners[id][name]
```

**Comment by codynova** [comment]
Cool that's exactly what I had imagined. Didn't mean to imply we'd use the RXJS lib, I just want to make sure I understand the pattern properly :smile:

**Comment by codynova** [comment]
New api which allows calling `.subscribe()` for body props (#57) released in 0.4.0, closing
